shared_model/builders/protobuf/builder_templates/query_response_template.hpp

```diff
@@ -0,0 +1,270 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2018 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_QUERY_RESPONSE_BUILDER_TEMPLATE_HPP
+#define IROHA_PROTO_QUERY_RESPONSE_BUILDER_TEMPLATE_HPP
+
+#include "backend/protobuf/query_responses/proto_query_response.hpp"
+#include "builders/protobuf/helpers.hpp"
+#include "common/visitor.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "responses.pb.h"
+
+namespace shared_model {
+  namespace proto {
+
+    template <class T>
+    static void setReason(iroha::protocol::ErrorResponse &err);
```

Comment:
Consider reworking this function, instead of setting a variable, use it as a map from type to enum and set error separately, so:
`template <typename T> constexpr ErrorResponse getReason()`

and in builder:
`query_response.set_reason(getReason<T>())`

This is cleaner from a functional point of view. Also, it allows declaring getReason as constexpr, which means that function call will be automatically inlined, which is always nice :)

See proof of inlining:
https://godbolt.org/g/XiZjb2